xclim/indices/stats.py

```diff
@@ -153,6 +155,62 @@ def fit(
     return out
 
 
+def fit_group(
+    da: xr.DataArray,
+    dist: str = "gamma",
+    freq: str = "MS",
+    window: int = 1,
+) -> xr.DataArray:
+    """Fit a resampled array averaged over a rolling window to a univariate distribution along the time dimension.
+
+    Parameters
+    ----------
+    da : xarray.DataArray
+      Input data.
+    dist : str
+      Name of the univariate distribution, such as `beta`, `expon`, `genextreme`, `gamma`, `gumbel_r`, `lognorm`, `norm`
+      (see scipy.stats).
+    freq : str
+      Resampling frequency.
+    window : int
+      Averaging window length (In units of the resampling frequency period).
+
+
+    Returns
+    -------
+    xr.DataArray
+      An array of fitted distribution parameters for each time segment
+
+
+    Notes
+    -----
+    The resampling frequency time delta must be an integer number of months. It determines time segments used to fit
+    the array group by group. For example, choosing `freq = 'MS` will give 12 time segment, i.e. months.
+    A time series for each month is used to obtain a set of fitting parameters.
+    """
+    # Freq must have a time delta which is an integer number of month
+    dts = pd.date_range("1950-01-01", periods=2, freq=freq).to_pydatetime()
+    if dts[-1].day - dts[0].day != 0 or dts[-1].second - dts[0].second != 0:
+        raise NotImplementedError(f"Resampling frequency `{freq}` not supported.")
+
+    # groupby argument, depends on the smallest time period allowed with freq
+    group = "time.month"
+
+    # Resampling and rolling
+    da = da.resample(time=freq).mean()
```

Comment:
I understand that this is what SPI does. but I think I wouldn't expect a function named `fit_group` to average over the group.  When I first suggested this, I thought the SPI fit was taken over _all_ available (daily) values. Seems strange to me to actively remove information before fitting, but that's ok for SPI if that's how they do it.

But, I think a `fit_group` function should fit on the group, not average the group, so the mean could be taken in the `spi` function, before calling this. However, this will break your "window" logic...

Reply:
Rather, my question is on why we couldn't have groups like:
group1 = {Jan 1 yr1, Jan 2 yr 1, Jan 31 yr1, Jan 1 yr2,... Jan 31 yr2, ....}
and so one.

Here the function "fit over the averaged group" instead of "fit over the group". I guess these indices have strict definitions, but wouldn't an SPI that was calibrated on all values actually yield a better results? since, there were more points involved in the fitting?

All that said, if we put the resampling outside this function, the rolling will break, unless we express the window in terms of timesteps of the input data instead of groups.

I think we need to discuss this further, as I may not understand the needs of SPI...

Reply:
For the behaviour of the `window` argument, I initially thought it would look like what `xclim.sdba.Grouper` does : if the data is daily, window=31 and group='month', then we have:
group1 = {Dec 15 yr0, ... Feb 15 yr1, Dec 15 yr1, ..., Feb 15 yr2}
group2 = {Jan 15 yr1, ..., Mar 15 yr1, Jan15 yr2, ..., Mar 15 yr2}
and so on

This is different from what is done here and thus from what SPI wants. But I believe it is what would make sense in a `fit_group` function. Meaning that we could simply downstream your resampling and rolling to the SPI indice?

Reply:
Reply to first comment
------------------------------
I'm not sure what is better from a statistical point of view. Fitting:
 group1 = {Jan 1 yr1, Jan 2 yr 1, Jan 31 yr1, Jan 1 yr2,... Jan 31 yr2, ....}
 would indeed give more data to work with, but you would also work with more variability in the dataset, no?
 
 As for definition of SPI, I think indeed it is often applied on a "monthly" dataset, which must mean you don't have information about specific days. For instance: 
 https://www.in.gov/dnr/water/water-availability-use-rights/water-resource-updates/monthly-water-resource-summary/explanation-of-standard-precipitation-index-spi/
 "A 1-month SPI typically compares well to the percent of normal precipitation for the 
  
 Reply to second comment
 -----------------------------------
 Ok, I see the difference. However, in what you describe "data is daily, window=31 and group='month'," I don't see how you would know that DEC,JAN,FEB are grouped together, then JAN,FEB,MAR, then ...  etc

Yes, we can put resampling and rolling directly in the SPI indice, and then simply call `fit` from within that call, removing the need for fit_group.

Reply:
Ok, having read your link and some other description, I really think  the resampling and rolling should be specific to SPI. 
If you can adapt `fit` so that `fit(da.groupby('time.month'), ...)` is possible, then maybe that's the simplest solution! Otherwise, we can keep a very simple `fit_group`.

Reply:
Yes `fit_group` can be avoided very easily. Only a few lines are necessary to add in the SPI code

```python
pr_cal = pr_cal.resample(time=freq).
pr_cal = pr_cal.rolling(time=window).mean(skipna=False, keep_attrs=True)
group = "time.month"
params = pr_cal.groupby(group).map(lambda x: fit(x, dist, method, dim))
```

By the way, we could give the possibility of a daily SPI. We would need to allow Higher resampling frequency (lower period, for instance `freq="D"`) and choose an appropriate `group = "time.dayofyear"` or something like this given this frequency.